File: abc201-220/abc203/e/main.py
from collections import defaultdict

N, M = map(int, input().split())
black = defaultdict(set)
for i in range(M):
	x, y = map(int, input().split())
	black[x].add(y)
dp = set([N])
for i in sorted(black.keys()):
	dpadd = set()
	dprem = set()
	for j in black[i]:
		if j not in dp and (j - 1 in dp or j + 1 in dp):
			dpadd.add(j)
		if j in dp and (j - 1 not in dp and j + 1 not in dp):
			dprem.add(j)
	for j in dpadd:
		dp.add(j)
	for j in dprem:
		dp.remove(j)
print(len(dp))

File: abc201-220/abc203/e/main.cs
using System;
using System.Linq;
using System.Collections.Generic;

class Program
{
	static void Main(string[] args)
	{
		var NM = Console.ReadLine().Split().Select(int.Parse).ToArray();
		var N = NM[0];
		var M = NM[1];
		var BlackPawns = new Dictionary<int, HashSet<int>>();
		for (int i = 0; i < M; i++)
		{
			var XY = Console.ReadLine().Split().Select(int.Parse).ToArray();
			var X = XY[0];
			var Y = XY[1];
			if (!BlackPawns.ContainsKey(X))
			{
				BlackPawns.Add(X, new HashSet<int>());
			}
			BlackPawns[X].Add(Y);
		}

		var dp = new HashSet<int>();
		dp.Add(N);
		foreach (var (_, bk) in BlackPawns.OrderBy(x => x.Key))
		{
			var dpAdd = new List<int>();
			var dpRem = new List<int>();
			foreach (var j in bk)
			{
				if (!dp.Contains(j) && (dp.Contains(j - 1) || dp.Contains(j + 1)))
				{
					dpAdd.Add(j);
				}
				if (dp.Contains(j) && (!dp.Contains(j - 1) && !dp.Contains(j + 1)))
				{
					dpRem.Add(j);
				}
			}
			foreach (var j in dpAdd) dp.Add(j);
			foreach (var j in dpRem) dp.Remove(j);
		}

		Console.WriteLine(dp.Count);
	}
}
